analytics/src/main/java/com/segment/analytics/internal/AnalyticsClient.java

```diff
@@ -214,7 +248,9 @@ public Looper() {
 
     @Override
     public void run() {
-      List<Message> messages = new ArrayList<>();
+      LinkedList<Message> messages = new LinkedList<>();
+      Map<String, Integer> messageIdSizeMap = new HashMap<>();
```

Comment:
we don't need a map here. use a `AtomicInteger` instead to accumulate the total size so far, and do batch based on it. using a hashmap here is a waste of memory, and it's not thread safe, since it's possible that `put` and `remove` happens at the same time, thus causing `ConcurrentModificationException`. `ConcurrentHashMap` could resolve the issue, but I would not recommend here, since it incurs a lot more computation and memory overhead then just using an atomic integer.

Reply:
@wenxi-zeng  We can calculate it here too but an AtomicInteger by itself is not gonna solve it, we still have to calculate it like this:

![image](https://user-images.githubusercontent.com/92827184/151896882-07b70000-0933-49c8-93c8-da36b5aa2c5c.png)


We can get rid of the map that way

Reply:
yep. that'd work

Reply:
hi @wenxi-zeng  can you review when you get a chance please? just committed the changes we discussed

Thanks!

Reply:
@jorgen25 the changes look good to me. please clear the CI test failures, then it's good to go.

Reply:
@wenxi-zeng  done! thanks a lot for your help btw!

Reply:
LGTM now. thanks for making these changes!